agent/consul/state/schema_test.go

```diff
@@ -54,42 +61,61 @@ func TestNewDBSchema_Indexers(t *testing.T) {
 					if !ok {
 						t.Skip("TODO: missing test case")
 					}
+					tc.run(t, indexer)
+				})
+			}
+		})
+	}
+}
 
-					args := []interface{}{tc.read.source}
-					if s, ok := tc.read.source.([]interface{}); ok {
-						// Indexes using memdb.CompoundIndex must be expanded to multiple args
-						args = s
-					}
+func (tc indexerTestCase) run(t *testing.T, indexer memdb.Indexer) {
+	args := []interface{}{tc.read.source}
+	if s, ok := tc.read.source.([]interface{}); ok {
+		// Indexes using memdb.CompoundIndex must be expanded to multiple args
+		args = s
+	}
 
-					actual, err := indexer.FromArgs(args...)
-					require.NoError(t, err)
-					require.Equal(t, tc.read.expected, actual)
+	if tc.read.source != nil {
+		t.Run("readIndex", func(t *testing.T) {
+			actual, err := indexer.FromArgs(args...)
+			require.NoError(t, err)
+			require.Equal(t, tc.read.expected, actual)
+		})
+	}
 
-					if i, ok := indexer.(memdb.SingleIndexer); ok {
-						valid, actual, err := i.FromObject(tc.write.source)
-						require.NoError(t, err)
-						require.True(t, valid)
-						require.Equal(t, tc.write.expected, actual)
-					}
+	if i, ok := indexer.(memdb.SingleIndexer); ok {
+		t.Run("writeIndex", func(t *testing.T) {
+			valid, actual, err := i.FromObject(tc.write.source)
+			require.NoError(t, err)
+			if tc.write.expectedIndexMissing {
+				require.False(t, valid, "expected the indexer to produce no index value")
+			} else {
+				require.True(t, valid, "indexer was missing a required value")
+				require.Equal(t, tc.write.expected, actual)
+			}
+		})
+	}
 
-					if i, ok := indexer.(memdb.PrefixIndexer); ok {
-						for _, c := range tc.prefix {
-							t.Run("", func(t *testing.T) {
-								actual, err := i.PrefixFromArgs(c.source)
-								require.NoError(t, err)
-								require.Equal(t, c.expected, actual)
-							})
-						}
-					}
+	if i, ok := indexer.(memdb.PrefixIndexer); ok {
+		for _, c := range tc.prefix {
+			t.Run("prefixIndex", func(t *testing.T) {
+				actual, err := i.PrefixFromArgs(c.source)
+				require.NoError(t, err)
+				require.Equal(t, c.expected, actual)
+			})
+		}
+	}
 
-					if i, ok := indexer.(memdb.MultiIndexer); ok {
-						valid, actual, err := i.FromObject(tc.writeMulti.source)
-						require.NoError(t, err)
-						require.True(t, valid)
-						require.Equal(t, tc.writeMulti.expected, actual)
-					}
-				})
-			}
+	if i, ok := indexer.(memdb.MultiIndexer); ok {
+		valid, actual, err := i.FromObject(tc.writeMulti.source)
+		require.NoError(t, err)
+		require.True(t, valid)
+		require.Equal(t, tc.writeMulti.expected, actual)
+	}
+
+	for _, extra := range tc.extra {
+		t.Run("extra", func(t *testing.T) {
```

Comment:
Does it matter that these all end up with the same name? A cheap fix would be to include the slice index in the name like `extra[0], extra[1]` etc?

Reply:
This is exactly what Go does automatically for us! If there are duplicate names it adds `#01` , `#02`, etc.